### PR TITLE
DEV: Support excluding categories with the `category:` filter

### PR DESCRIPTION
**Pull request by tgxworld**
This commit adds support for excluding categories when using the
`category:` filter with the `-` prefix. For example,
`-category:category-slug` will exclude all topics that belong to the
category with slug "category-slug" and all of its sub-categories.

To only exclude a particular category and not all of its sub-categories,
the `-` prefix can be used with the `=` prefix. For example,
`-=category:category-slug` will only exclude topics that belong to the
category with slug "category-slug". Topics in the sub-categories of
"category-slug" will still be included.

### CODE REVIEW

**Comment by tgxworld** [comment]
@lis2 or @OsamaSayegh Can you help me to review this? Thank you!

File: lib/topics_filter.rb [hunk 1]
```diff
@@ -18,7 +18,7 @@ def filter_from_query_string(query_string)
     filters = {}
 
     query_string.scan(
-      /(?<key_prefix>[-=])?(?<key>[\w-]+):(?<value>[^\s]+)/,
+      /(?<key_prefix>(?:-|=|-=|=-))?(?<key>[\w-]+):(?<value>[^\s]+)/,
```
Comment (lis2) [inline lib/topics_filter.rb:21]: Is it right? My understanding is:
`-` - exclude category and subcategories
`=` - include exact category
`-=` - exclude exact category
`=-` - ?

File: lib/topics_filter.rb [hunk 2]
```diff
@@ -186,51 +186,80 @@ def filter_by_number_of_views(min: nil, max: nil)
...
+      exclude: {
+        with_subcategories: [],
+        without_subcategories: [],
+      },
+    }
 
     values.each do |key_prefix, value|
-      break if key_prefix && key_prefix != "="
+      exclude_categories = key_prefix&.include?("-") || false
+      exclude_subcategories = key_prefix&.include?("=") || false
```
Comment (lis2) [inline lib/topics_filter.rb:203]:
```suggestion
      exclude_categories = key_prefix&.include?("-")
      exclude_subcategories = key_prefix&.include?("=")
```

File: lib/topics_filter.rb [hunk 3]
```diff
@@ -186,51 +186,80 @@ def filter_by_number_of_views(min: nil, max: nil)
...
```
Comment (lis2) [inline lib/topics_filter.rb:202]:
```suggestion
      exclude_categories = key_prefix&.include?("-")
      exclude_subcategories = key_prefix&.include?("=")
```

File: lib/topics_filter.rb (hunk 1 above)
Comment (tgxworld) [reply lib/topics_filter.rb:21]: `-=` and `=-` does the sam thing :)

**Review by lis2 — APPROVED** [review]

**Comment by tgxworld** [comment]
Thank you @lis2

**Pull request merged**